src/small_dense_hash_map.hpp

```diff
@@ -0,0 +1,72 @@
+/*
+  Copyright (c) 2014-2016 DataStax
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+#ifndef __CASS_SMALL_DENSE_HASH_MAP_HPP_INCLUDED__
+#define __CASS_SMALL_DENSE_HASH_MAP_HPP_INCLUDED__
+
+#include "fixed_allocator.hpp"
+#include "macros.hpp"
+
+#include <assert.h>
+#include <sparsehash/dense_hash_map>
+
+// This should be the same as sparsehash::densehashtable<>::HT_OCCUPANCY_PCT
+// which can be found at the bottom of "sparesehash/internal/densehashtable.h".
+#define OCCUPANCY_PCT 50
+
+#define MIN_BUCKETS(N) STATIC_NEXT_POW_2((((N * 100) / OCCUPANCY_PCT) + 1))
```

Comment:
This is to statically allocate enough buckets to hold N items without requiring additional allocations. A hash table doesn't just allocate N buckets to hold N items because of hash collisions so it allocates a number of buckets over the current number of items (controlled by a load factor) to compensate for that.

`OCCUPANCY_PCT` is `dense_hash_map`'s load factor and it needs the number of buckets to be a power of two because it uses  a mask and `&` to wrap around instead of using modulo (which is expensive).

This calculates the number of buckets considering the load factor: `(N * 100) / OCCUPANCY_PCT ` (The `+ 1` is to allocate at least one bucket) then it rounds up to the next power of two for the wrap around trick.